pkg/package-server/storage/subresources.go

```diff
@@ -0,0 +1,92 @@
+package storage
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericreq "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider"
+)
+
+// LogoStorage implements Kubernetes methods needed to provide the `packagemanifests/icon` subresource
+type LogoStorage struct {
+	groupResource schema.GroupResource
+	prov          provider.PackageManifestProvider
+}
+
+var _ rest.Connecter = &LogoStorage{}
+var _ rest.StorageMetadata = &LogoStorage{}
+
+// NewLogoStorage returns struct which implements Kubernetes methods needed to provide the `packagemanifests/icon` subresource
+func NewLogoStorage(groupResource schema.GroupResource, prov provider.PackageManifestProvider) *LogoStorage {
+	return &LogoStorage{groupResource, prov}
+}
+
+// New satisfies the Storage interface
+func (s *LogoStorage) New() runtime.Object {
+	return &operators.PackageManifest{}
+}
+
+// Connect satisfies the Connector interface and returns the image icon file for a given `PackageManifest`
+func (s *LogoStorage) Connect(ctx context.Context, name string, options runtime.Object, responder rest.Responder) (http.Handler, error) {
+	var handler http.HandlerFunc = func(w http.ResponseWriter, r *http.Request) {
+		if match := r.Header.Get("If-None-Match"); match != "" && r.URL.Query().Get("resourceVersion") != "" {
+			w.WriteHeader(http.StatusNotModified)
+			return
+		}
+
+		namespace := genericreq.NamespaceValue(ctx)
+		pkg, err := s.prov.Get(namespace, name)
+		if err != nil || pkg == nil || len(pkg.Status.Channels[0].CurrentCSVDesc.Icon) == 0 {
```

Comment:
can we make this section more defensive? don't assume there are any channels, don't assume that Icon slice is not nil, etc.

Reply:
Added check for no channels, but `len(nil)` will return 0 in Go.